File: adventofcode-2024/src/main/java/nl/mvdr/adventofcode/adventofcode2024/day04/Part1.java
package nl.mvdr.adventofcode.adventofcode2024.day04;

import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.solver.LongSolver;

public class Part1 implements LongSolver {

    private static final Logger LOGGER = LoggerFactory.getLogger(Part1.class);

    @Override
    public long solve(Stream<String> lines) {
        return WordSearch.parse(lines.toList()).countXmasMatches();
    }

    public static void main(String[] args) {
        var instance = new Part1();

        var result = instance.solve("input-day04-2024.txt");

        LOGGER.info(result);
    }
}
 

File: adventofcode-2025/src/main/kotlin/nl/mvdr/adventofcode/adventofcode2025/day04/Part1.kt
package nl.mvdr.adventofcode.adventofcode2025.day04

import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.point.CharacterConsumer
import nl.mvdr.adventofcode.point.Point
import nl.mvdr.adventofcode.solver.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart1(lines: Sequence<String>): Int {
    val rolls = parseInput(lines)
    return rolls.count { roll -> roll.neighboursIncludingDiagonals().count(rolls::contains) < 4 }
}

private fun parseInput(lines: Sequence<String>): Set<Point> {
    val result = mutableSetOf<Point>();
    Point.parse2DMap(lines.toList(), CharacterConsumer { point, character -> if (character == '@') { result.add(point) } })
    return result
}

fun main() {
    val result = FunctionSolver(::solvePart1).solve("input-day04.txt")
    logger.info { result }
}
